Convert to Flask App

robputt opened the pull request.

marvinmarnold commented on hw_diag/utilities/hardware.py:
```diff
@@ -0,0 +1,118 @@
+    Returns true or false.
+    """
+    result = None
+    while result is None:
```
Couldn't this get stuck in an infinite loop if pktfwd never loads?

marvinmarnold replied:
Can we use Python's native retry logic? https://julien.danjou.info/python-retrying/

robputt replied:
It needs to load eventually, else. the miner is essentially totally broken, so if we wait forever or if we retry it. is kind of irrelevant at that point.

marvinmarnold replied:
In that case, wouldn't we still want to display the diagnostics page but show that there's a serious error? The way it is now, the program will block forever.

robputt replied:
Handle as bug seperately.

marvinmarnold commented on hw_diag/utilities/miner.py:
```diff
@@ -0,0 +1,77 @@
+            param_list = [
+                str(p2pstatus[0][1]),
+                str(p2pstatus[1][1]),
+                str(p2pstatus[3][1]),
```
Just checking that this ordering is intentional. Is [this](https://github.com/helium/miner/blob/master/src/miner.erl#L175) the right place to look for the source data format?

marvinmarnold commented on hw_diag/utilities/shell.py:
```diff
@@ -0,0 +1,41 @@
+    ]
+    keys = ["BN", "ID", "BA", "FR", "FW", "VA"]
+
+    for (var, key) in zip(env_var, keys):
```
We should probably create a dictionary and iterate over that

env_var_mappings = {
  'BALENA_DEVICE_NAME_AT_INIT': 'BN',
  ...
}

But really, why not just use the env var name as the key on diagnostics? I made the same comment earlier. Is there a benefit in renaming the keys to be shorter?

robputt replied:
This code is lifted from the rework that Ilya completed previously. Not sure if there is any benefit in using a hash table over using 2 lists and the zip function, I didn't really feel the need to change it.

We cannot alter the keys as we need to retain backwards compatibility, a requirement from Viktor, as the factory tooling relies on this naming convention. Renaming these keys should be completed though, as they are ugly right now, but this is probably out of scope for a Flask conversion and needs to be a story for the future so we can ensure the factory tooling is updated.

marvinmarnold commented on hw_diag/utilities/variant_definitions.py:
```diff
@@ -0,0 +1,215 @@
+variant_definitions = {
```
This is being copied from [another library](https://raw.githubusercontent.com/NebraLtd/helium-hardware-definitions/master/variant_definitions.py). Following Flask conventions, is this the right place? I've been putting this in a lib/` folder.

robputt replied:
Flask is mostly unopinionated about this kind of stuff, really this should be packaged and included via requirements.txt so it is easily re-usable in other projects, but thats out of scope for this story to convert the existing logic to Flask. I would suggest we create a story to find a better way on how to distribute and use this file.

marvinmarnold replied:
Agreed: https://github.com/NebraLtd/helium-hardware-definitions/issues/1

marvinmarnold commented on hw_diag/views/diagnostics.py:
```diff
@@ -0,0 +1,53 @@
+    if diagnostics.get('error'):
+        return 'Internal Server Error', 500
+
+    response = {
```
Why do we return a subset of all information?

robputt replied:
A requirement was to retain backwards compatibility with the old app, this is used to generate /initFile.txt and is used in the factory for device onboarding.

marvinmarnold reviewed (dismissed):
I left a lot of comments but this is a big improvement over what's there. I know many of my comments address pre-existing issues. We don't need to address all of the things but I think we can improve readability a bit. At least a few tests should also be added.

robputt commented on hw_diag/utilities/hardware.py:
```diff
@@ -0,0 +1,118 @@
+            logging.error(e)
+
+
+def get_ethernet_addresses(diagnostics):
```
Yep this is a historic thing, remember https://github.com/NebraLtd/hm-diag/issues/60 is all about converting this to a Flask app, not a larger refactor of other elements of the code base. These files are largely moved from the previous implementation. We should split reworking this into a separate story.

marvinmarnold replied:
Added a requirement to rename all other methods consistently in https://github.com/NebraLtd/hm-diag/issues/95#issue-961891227

robputt commented on hw_diag/utilities/hardware.py:
```diff
@@ -0,0 +1,118 @@
+from hw_diag.utilities.shell import config_search_param
+
+
+def detect_bt_lte(diagnostics):
```
Renamed...

robputt commented on hw_diag/utilities/blockchain.py:
```diff
@@ -0,0 +1,23 @@
+            )
+        return result
+    else:
+        return "1"
```
Now returns None on error.

robputt commented on hw_diag/templates/diagnostics_page.html:
```diff
@@ -0,0 +1,152 @@
+                  {% endif %}
+                </tr>
+                <tr>
+                  <td>LoRa OK?</td>
```
See https://github.com/NebraLtd/hm-diag/issues/61

robputt commented on hw_diag/tasks.py:
```diff
@@ -0,0 +1,110 @@
+        diagnostics['MR'] = False
+
+    # Get the blockchain height from the Helium API
+    value = "1"
```
Unsure, this is historic code, if in doubt leave it in so it doesn't break stuff, prefer not to perform scream tests.

robputt commented on hw_diag/tasks.py:
```diff
@@ -0,0 +1,110 @@
+    try:
+        data = get_public_keys()
+    except PermissionError as e:
+        data = None
```
With it outside of the try block None still gets written to the file. With it in the try block the file does not get written.

robputt commented on hw_diag/tasks.py:
```diff
@@ -0,0 +1,110 @@
+    get_rpi_serial(diagnostics)
+    detect_ecc(diagnostics)
+
+    diagnostics['LOR'] = lora_module_test()
```
Nope, we must retain this structure as the factory tooling uses it :-(.

robputt commented on hw_diag/app.py:
```diff
@@ -0,0 +1,51 @@
+
+
+if __name__ == '__main__':
+    main()
```
Removed

robputt commented on hw_diag/tasks.py:
```diff
@@ -0,0 +1,110 @@
+log.setLevel(logging.DEBUG)
+
+
+def perform_hw_diagnostics():
```
In my opinion out of scope for this PR, this function is lifted from previous version and wrapped in APScheduler. The aim of this PR is to convert to Flask not massively refactor the other business logic. We should tackle this in a separate story.

marvinmarnold reviewed (commented):
Looking real good with all those changes. I'd still like to see some tests added as part of this work. If you don't think you've added any net new logic that needs to be tested, then could you add a handful of tests for pre-existing logic to establish a pattern for how Flask tests should be done?

Could you also add some documentation? Should include how to run tests. Also if there is a recommended way to run the site locally during development.

shawaj commented on hw_diag/utilities/hardware.py:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    for dev_type, dev_addr in devices.items():
+        resp = os.popen('grep 0a12 /sys/bus/usb/devices/*/idVendor').read()
```
shouldn't this have a variable instead of just `0a12` after the `grep`?

As currently it won't find the LTE will it?

robputt replied:
Thanks @shawaj nice catch I'll get this sorted.

robputt commented on hw_diag/tasks.py:
```diff
@@ -36,11 +36,10 @@ def perform_hw_diagnostics():
 
```
https://github.com/NebraLtd/Hotspot-Production-Tool/blob/main/main.py#L199

shawaj commented on hw_diag/utilities/variant_definitions.py:
```diff
@@ -0,0 +1,215 @@
+        'RESET': 38,
+        'MAC': 'eth0',
+        'STATUS': 25,
+        'BUTTON': 26,
```
BUTTON needs changing to 24

See https://github.com/NebraLtd/hm-pktfwd/pull/31

shawaj replied:
Now packaged Enjoy https://pypi.org/project/hm-hardware-defs/

You want to use v0.1.4

See also https://github.com/NebraLtd/hm-pktfwd/pull/32

shawaj commented on hw_diag/utilities/variant_definitions.py:
```diff
@@ -0,0 +1,215 @@
+        'RESET': 38,
+        'MAC': 'eth0',
+        'STATUS': 25,
+        'BUTTON': 26,
```
BUTTON needs changing to 24

See https://github.com/NebraLtd/hm-pktfwd/pull/31

shawaj replied:
Now packaged Enjoy https://pypi.org/project/hm-hardware-defs/

You want to use v0.1.4

See also https://github.com/NebraLtd/hm-pktfwd/pull/32

shawaj commented:
Looks good to me @robputt - tested it on a couple of test devices and works nicely.

Few things I noticed, but pretty minor and can probably be fixed later I guess:
- on the backend logs it was complaining about not having a timezone set and saying it was a development server and not to use in production
- some of the font downloads were 404ing according to logs (nunito-regular.woff and nunito-bold.woff)
- when the region has been set with a REGION_OVERRIDE env variable instead of directly via the miner it gives a "region plan" of UN123 instead of what's in the region override. Not sure if this makes sense or not. For production we don't use an override but in testing of packet forwarder and light hotspots we sometimes do (especially with non location asserted devices) see [here in packet forwarder](https://github.com/NebraLtd/hm-pktfwd/blob/c12a73944d23bba92e57cdd87043d56fe0b37819/files/configurePktFwd.py#L47). Perhaps if there is a region override set, we should mention that in the diagnostics like `EU868 (Set by override)` and could do similar when set by miner perhaps.
- in the original diag, instead of showing UN123 directly it showed ["awaiting location assertion"](https://github.com/NebraLtd/hm-diag/blob/99f83a011bf1c016adcd6a7cc7272573f5925612/diagnostics-program/html_generator.py#L126)
- LTE detected shows even for non-outdoor units. I'm wondering whether we should add a field to the [hardware variants](https://github.com/NebraLtd/helium-hardware-definitions) of "LTE Support" or something to that effect? So that we can automatically determine whether it should even bother to check for LTE?
- on a side note, perhaps cellular detected or modem detected is better than LTE?
- could add a favicon perhaps?

Going to mark as approved anyway as I guess all this can be resolved in a later update.

shawaj commented:
Some screenshots of it running on a RAK based light hotspot I tested on. Looks smart!

![Screenshot_20210814-195622_Chrome](https://user-images.githubusercontent.com/3359418/129457597-a915a42a-4390-4e34-a557-7792f41d65f8.jpg)
![Screenshot_20210814-195630_Chrome](https://user-images.githubusercontent.com/3359418/129457600-15d1b266-7e6d-4377-901a-dea849ae121e.jpg)
![Screenshot_20210814-195639_Chrome](https://user-images.githubusercontent.com/3359418/129457601-a1496ef9-52db-4292-a5c8-ae82101d0ddd.jpg)

robputt commented:
> Looks good to me @robputt - tested it on a couple of test devices and works nicely.
> 
> Few things I noticed, but pretty minor and can probably be fixed later I guess:
> 
> * on the backend logs it was complaining about not having a timezone set and saying it was a development server and not to use in production
> * some of the font downloads were 404ing according to logs (nunito-regular.woff and nunito-bold.woff)
> * seems to download the font every time the page re-loads automatically and therefore isn't instantly applied. Could we just cache it locally instead or bundle the font or something?
> * when the region has been set with a REGION_OVERRIDE env variable instead of directly via the miner it gives a "region plan" of UN123 instead of what's in the region override. Not sure if this makes sense or not. For production we don't use an override but in testing of packet forwarder and light hotspots we sometimes do (especially with non location asserted devices) see [here in packet forwarder](https://github.com/NebraLtd/hm-pktfwd/blob/c12a73944d23bba92e57cdd87043d56fe0b37819/files/configurePktFwd.py#L47). Perhaps if there is a region override set, we should mention that in the diagnostics like `EU868 (Set by override)` and could do similar when set by miner perhaps.
> * in the original diag, instead of showing UN123 directly it showed ["awaiting location assertion"](https://github.com/NebraLtd/hm-diag/blob/99f83a011bf1c016adcd6a7cc7272573f5925612/diagnostics-program/html_generator.py#L126)
> * LTE detected shows even for non-outdoor units. I'm wondering whether we should add a field to the [hardware variants](https://github.com/NebraLtd/helium-hardware-definitions) of "LTE Support" or something to that effect? So that we can automatically determine whether it should even bother to check for LTE?
> * on a side note, perhaps cellular detected or modem detected is better than LTE?
> * could add a favicon perhaps?
> 
> Going to mark as approved anyway as I guess all this can be resolved in a later update.


* QUESTION - The TZ issue is a Dockerism as Docker containers do not have a time zone set usually and hence default to UTC. I personally like UTC for machines as it makes them easily comparable if they are all in 1 universal time zone, however if you prefer to set a time zone please let me know which TZ and we'll create an issue for future enhancement for TZ support. The current deployed version also defaults to UTC it just doesn't scream about it.
* QUESTION - Using development server - currently the container just runs the CherryPi Python HTTP server as I didn't feel the need to bloat with NGINX / Apache, however as per the warning it is not a production ready server, mostly in terms of scaling although this should be a moot point as only a few requests are every going to hit these endpoints and they *should* be internal to the customers network. Although if you prefer to use WSGI and use a "real" web server I can edit the docker accordingly.
* NEW ISSUE -  Add caching headers for static files to prevent re-download each page refresh -  see https://github.com/NebraLtd/hm-diag/issues/102
* NEW ISSUE -  Favicon - see https://github.com/NebraLtd/hm-diag/issues/100
* FIXED - LTE detected now says "Modem Detected" in the template as requested
* NEW ISSUE - LTE showing on both indoor and outdoor variants - see https://github.com/NebraLtd/hm-diag/issues/101 (this is a regression from the previous version although I believe it should be handled separately as this has now been added to the variants file and will require some rework.
* FIXED - UN123 should show "Awaiting Location Assertion"
